### PR TITLE
Extendable discovery configuration

### LINKED ISSUE #119: Cannot customize ConsulLifecycle
I noticed two problems when trying to use a subclass of `ConsulLifecyce`:
- in multiple configuration classes injection of dependencies is not really injecting them, but creating a separate copy (because it's attempted via a method call on the same class)
- there is no ConditionalOnMissingBean on ConsulLifecycle. Marking second implementation as `@Primary` starts both lifecycles and overriding beans by name is not the best way.

### PR DESCRIPTION
**Pull request by mabn**
Some fixes/refactoring in autoconfigurations which make providing own implementations of certain beans possible (in particular ConsulLifecycle)

Fixes #119

### CODE REVIEW

File: spring-cloud-consul-config/src/main/java/org/springframework/cloud/consul/config/ConsulConfigBootstrapConfiguration.java [hunk 1]
```diff
@@ -34,21 +34,18 @@
 public class ConsulConfigBootstrapConfiguration {
 
 	@Configuration
-	@EnableConfigurationProperties
+	@EnableConfigurationProperties(ConsulConfigProperties.class)
```
Comment (spencergibb) [inline spring-cloud-consul-config/src/main/java/org/springframework/cloud/consul/config/ConsulConfigBootstrapConfiguration.java]: I'd rather stick to the way we were creating the properties, please.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClientConfiguration.java [hunk 2]
```diff
@@ -32,45 +33,42 @@
 @Configuration
 @ConditionalOnConsulEnabled
 @ConditionalOnProperty(value = "spring.cloud.consul.discovery.enabled", matchIfMissing = true)
-@EnableConfigurationProperties
+@EnableConfigurationProperties({LifecycleProperties.class, ConsulDiscoveryProperties.class, HeartbeatProperties.class})
 public class ConsulDiscoveryClientConfiguration {
 
 	@Autowired
 	private ConsulClient consulClient;
+	@Autowired
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClientConfiguration.java]: Don't autowire beans created in this class, pass them as parameters to the method.

**Comment by spencergibb** [comment]
@mabn any response to my comments? If not, I'm going to close this.

**Comment by mabn** [comment]
@spencergibb done

File: spring-cloud-consul-bus/src/main/java/org/springframework/cloud/consul/bus/ConsulBusAutoConfiguration.java [hunk 3]
```diff
@@ -68,16 +69,17 @@ public ConsulOutboundEndpoint consulOutboundEndpoint() {
 
 	@Bean
 	public IntegrationFlow cloudBusConsulOutboundFlow(
-			@Qualifier("cloudBusOutboundChannel") MessageChannel cloudBusOutboundChannel) {
+			@Qualifier("cloudBusOutboundChannel") MessageChannel cloudBusOutboundChannel,
```
Comment (spencergibb) [inline spring-cloud-consul-bus/src/main/java/org/springframework/cloud/consul/bus/ConsulBusAutoConfiguration.java]: Why did you touch files outside of discovery?

Comment (mabn) [inline spring-cloud-consul-bus/src/main/java/org/springframework/cloud/consul/bus/ConsulBusAutoConfiguration.java]: As I mentioned in #119 
> in multiple configuration classes injection of dependencies is not really injecting them, but creating a separate copy (because it's attempted via a method call on the same class)

so i fixed all the occurrences I was able to find

Comment (spencergibb) [inline spring-cloud-consul-bus/src/main/java/org/springframework/cloud/consul/bus/ConsulBusAutoConfiguration.java]: Please limit your changes to discovery. This is out of scope for this PR.

**Comment by spencergibb** [comment]
Squashed and merged with a little polish via ccea77d6615a22783daecbede15deb12dc6cef04
